### PR TITLE
Parallelize project builds

### PR DESCRIPTION
**Pull request by justice-adams-apple**
### Pull Request Description

Parallelize project builds to speed up testing time significantly

As part of this PR
* Converted the factory pattern to use a `FactoryBuilder` object for handling object instantioation. This allows the factory pattern to be pickle-able which is required for multiprocessing support
* Implemented the `ProjectListBuilder.build()` function in order to parallelize the project builds across the desired number of processes
* Cleaned up some of the `XcodeTarget` logic which was printing various messages to stdout when it should have been printing to the log files. This became increasingly obvious/annoying after adding multiprocessing support

### **Factory work**
Previously we had a factory which basically encapsulated certain constructor args with a lambda. However, this is un-pickable from python standards which means none of those objects could be passed down to the multiprocessing support. From "[what can be pickled](https://docs.python.org/3/library/pickle.html#what-can-be-pickled-and-unpickled)"
> functions (built-in and user-defined) accessible from the top level of a module (using [def](https://docs.python.org/3/reference/compound_stmts.html#def), not [lambda](https://docs.python.org/3/reference/expressions.html#lambda));

Thus, I had to create an actual well defined object to house the constructor args which are used by the factory method. This makes the objects pickle-able and opens the door for multiprocessing support. This is the motivation behind `FactoryBuilder`

### **Multiprocessing**
We need to parallelize at the project level as to avoid conflicting git operations (you can't maintain two revisions of a git repo without maintaining two repositories), thus I implemented `ProjectListBuilder.build()` and was able to easily plug in the `concurrent.futures` library to do the heavy lifting with respect to the multi-processing of the project builds. I've included a configurable `--process-count` argument the end-user can override

### CODE REVIEW

**Comment by justice-adams-apple** [comment]
@swift-ci test with toolchain

**Comment by justice-adams-apple** [comment]
@swift-ci test

**Review by shahmishal — APPROVED** [review]
Thanks!

**Comment by justice-adams-apple** [comment]
@swift-ci test

**Comment by justice-adams-apple** [comment]
@swift-ci test

**Comment by justice-adams-apple** [comment]
Local timings on my M1 (Just running the test suite):
* `main` -> 77m
* `PR/jadams-parallelize-projects` -> 15m


CI Timings :
* Previous PR: [4hr 33min](https://ci.swift.org/job/swift-PR-source-compat-suite-test-macOS/37)
* This PR: [2hr 4 min](https://ci.swift.org/job/swift-PR-source-compat-suite-test-macOS/38)

Waiting for green CI run before merging

**Pull request merged**
